### PR TITLE
CSCQVAIN-324: ADD: Error Handling

### PR DESCRIPTION
**Pull request by ShreyasDeshpande**
Error handling by showing error_id , returned from api error message. In other cases like axios, if request is getting timeout, then show error messaage

### CODE REVIEW

File: src/views/Datasets.vue [hunk 1]
```diff
@@ -431,17 +428,24 @@ const fields = [
 	},
 ]
 
-function getApiError(error) {
-	let apiError = "API Error"
+function getApiError(error,apiCall,datasetId) {
+	let errorText = "Error "
+	errorText+=apiCall
+	errorText+=datasetId
```
Comment (jppiiroinen) [inline src/views/Datasets.vue]: should it be something like:
let errorText = ["Error", apiCall, datasetId].join(" ")

File: src/views/Datasets.vue [hunk 2]
```diff
@@ -431,17 +428,24 @@ const fields = [
...
+			errorText += ": " + error.response.data.msg
 		}
-	} else if (error.message) {
-		apiError += ": " + error.message.toLowerCase()
+		if(error.response.data.error_id) {
+			errorText+=" Error id = "+error.response.data.error_id+" Please contact servicedesk(at)csc.fi. "
+		}
+	} else if (error.code  && error.code === 'ECONNABORTED') {
+		errorText += ": " + "Request is taking too long "
+	}else if (error.message) {
```
Comment (jppiiroinen) [inline src/views/Datasets.vue]: whitespace

File: src/views/Datasets.vue (hunk 1 above)
Comment (jppiiroinen) [reply src/views/Datasets.vue]: use similar kind of things below, having whitespace characters in the strings looks nasty

File: src/views/Editor.vue [hunk 3]
```diff
@@ -529,6 +538,28 @@ export default {
 				}
 			})
 		},
+		getError(error,apiCall,datasetId) {
```
Comment (jppiiroinen) [inline src/views/Editor.vue]: is this function a full copy?

**Review by jppiiroinen — CHANGES_REQUESTED** [review]

File: src/views/Editor.vue (hunk 3 above)
Comment (ShreyasDeshpande) [reply src/views/Editor.vue]: Slight change as Dataset handles both errors (backend and timeout etc). However for Editor we had to apply check

File: src/views/Datasets.vue (hunk 1 above)
Comment (ShreyasDeshpande) [reply src/views/Datasets.vue]: This is a good suggestion. applied!

Comment (ShreyasDeshpande) [reply src/views/Datasets.vue]: changed code

File: src/views/Datasets.vue (hunk 2 above)
Comment (ShreyasDeshpande) [reply src/views/Datasets.vue]: changed

File: src/views/Editor.vue (hunk 3 above)
Comment (ShreyasDeshpande) [reply src/views/Editor.vue]: modified existing method to have common body of both methods. and exported into a new file

File: src/views/Editor.vue [hunk 4]
```diff
@@ -403,10 +410,12 @@ export default {
 
 				const isExisting = (currentId && currentId !== 'new')
 				if (isExisting) {
+					errorMethod = " While updating dataset "
```
Comment (jppiiroinen) [inline src/views/Editor.vue]: " While updating dataset " => "While updating dataset"

File: src/views/Editor.vue [hunk 5]
```diff
@@ -392,6 +398,7 @@ export default {
 			}
 		},
 		save: async function saveCallback() {
+			let errorMethod = null
```
Comment (jppiiroinen) [inline src/views/Editor.vue]: errorMethod => errorDescription ?

File: src/views/Editor.vue [hunk 6]
```diff
@@ -381,6 +384,9 @@ export default {
 				if (e.response && e.response.status == 401) {
 					this.handleLostSession()
 				}
+				if(e.code ==='ECONNABORTED') {
+					this.errorMessage = getApiError(e," While Publishing dataset:",this.$store.state.metadata.id)
```
Comment (jppiiroinen) [inline src/views/Editor.vue]: getApiError(e," While Publishing dataset:",this.$store.state.metadata.id) =>
getApiError(e, "While Publishing dataset", this.$store.state.metadata.id)

File: src/views/Editor.vue [hunk 7]
```diff
@@ -403,10 +410,12 @@ export default {
 
 				const isExisting = (currentId && currentId !== 'new')
 				if (isExisting) {
+					errorMethod = " While updating dataset "
 					payload.id = currentId
 					await apiClient.put("/datasets/" + currentId, payload)
 					await this.openRecord(currentId)
 				} else {
+					errorMethod = " While saving dataset "
```
Comment (jppiiroinen) [inline src/views/Editor.vue]: " While saving dataset " => "While saving dataset"

File: src/views/Editor.vue [hunk 8]
```diff
@@ -418,6 +427,9 @@ export default {
 				if (error.response && error.response.status == 401) {
 					this.handleLostSession()
 				}
+				else {
+					this.errorMessage = getApiError(error,errorMethod,this.$store.state.metadata.id)
```
Comment (jppiiroinen) [inline src/views/Editor.vue]: getApiError(error,errorMethod,this.$store.state.metadata.id) => getApiError(error, errorMethod, this.$store.state.metadata.id)

File: src/views/Editor.vue [hunk 9]
```diff
@@ -498,9 +510,8 @@ export default {
 					this.handleLostSession()
 				}
 				else {
-					this.handleOtherError()
+					this.errorMessage = getApiError(error,"while opening dataset",id)
```
Comment (jppiiroinen) [inline src/views/Editor.vue]: getApiError(error,"while opening dataset",id) => getApiError(error, "while opening dataset", id)

File: src/views/Editor.vue [hunk 10]
```diff
@@ -304,8 +311,8 @@ export default {
 			qvainData: null,
 			reloadDatasetCounter: 0,
 			reloadDatasetTimer: null,
-			otherError: false,
 			openRecordCounter: 0,
+			errorMessage:null,
```
Comment (jppiiroinen) [inline src/views/Editor.vue]: errorMessage:null, => errorMessage: null,

File: src/views/Datasets.vue [hunk 11]
```diff
@@ -548,7 +531,7 @@ export default {
 					await this.$auth.logoutDueSessionTimeout()
 					this.$router.push({ name: "home", params: { missingSession: true }})
 				}
-				this.error = getApiError(e)
+				this.error = getApiError(e,"While deleting dataset",this.itemToBeDeleted.id)
```
Comment (jppiiroinen) [inline src/views/Datasets.vue]: getApiError(e,"While deleting dataset",this.itemToBeDeleted.id) => getApiError(e, "While deleting dataset", this.itemToBeDeleted.id)

File: src/views/Datasets.vue [hunk 12]
```diff
@@ -525,7 +508,7 @@ export default {
 					console.log("Show modal error")
 					this.$root.$emit('bv::show::modal', 'publishErrorModal')
 				} else {
-					this.error = getApiError(e)
+					this.error = getApiError(e,"While publishing dataset",this.itemToBePublished.id)
```
Comment (jppiiroinen) [inline src/views/Datasets.vue]: getApiError(e,"While publishing dataset",this.itemToBePublished.id) => getApiError(e, "While publishing dataset", this.itemToBePublished.id)

File: src/views/Datasets.vue [hunk 13]
```diff
@@ -497,7 +480,7 @@ export default {
 					await this.$auth.logoutDueSessionTimeout()
 					this.$router.push({ name: "home", params: { missingSession: true }})
 				}
-				this.error = getApiError(e)
+				this.error = getApiError(e,"While fetching datasets","")
```
Comment (jppiiroinen) [inline src/views/Datasets.vue]: getApiError(e,"While fetching datasets","") => getApiError(e, "While fetching datasets", "")

File: src/lib/getApiError.js [hunk 14]
```diff
@@ -0,0 +1,21 @@
...
+*/
+
+export default function getApiError(error,apiCall,datasetId) {
+	let errorText=[ "Error", apiCall, datasetId ]
+	if (error.response) {
+		errorText.push("[", error.response.status ,"]")
+		if (error.response.data && error.response.data.msg) {
+			errorText.push(":" , error.response.data.msg,",")
+		}
+		if(error.response.data.error_id) {
```
Comment (jppiiroinen) [inline src/lib/getApiError.js]: missing check for error.response.data

Comment (jppiiroinen) [inline src/lib/getApiError.js]: missing whitespace after if

File: src/lib/getApiError.js [hunk 15]
```diff
@@ -0,0 +1,21 @@
...
+			errorText.push("Error id =",error.response.data.error_id ,"Please contact servicedesk(at)csc.fi with this error Id")
```
Comment (jppiiroinen) [inline src/lib/getApiError.js]: errorText.push("Error id =",error.response.data.error_id ,"Please contact servicedesk(at)csc.fi with this error Id") => errorText.push("Error id =", error.response.data.error_id, "Please contact servicedesk(at)csc.fi with this error Id")

File: src/lib/getApiError.js [hunk 16]
```diff
@@ -0,0 +1,21 @@
...
+		}
+	} else if (error.code  && error.code === 'ECONNABORTED') {
+		errorText.push(":" , "Request is taking too long")
```
Comment (jppiiroinen) [inline src/lib/getApiError.js]: errorText.push(":" , "Request is taking too long") => errorText.push(":", "Request is taking too long")

File: src/lib/getApiError.js [hunk 17]
```diff
@@ -0,0 +1,21 @@
...
+	}else if (error.message) {
+		errorText.push(":", error.message.toLowerCase())
+	}
+	return errorText.join(" ")
+}
```
Comment (jppiiroinen) [inline src/lib/getApiError.js]: missing empty line

File: src/lib/getApiError.js [hunk 18]
```diff
@@ -0,0 +1,21 @@
+/*
+This function returns well texted error message
+*/
+
+export default function getApiError(error,apiCall,datasetId) {
+	let errorText=[ "Error", apiCall, datasetId ]
+	if (error.response) {
+		errorText.push("[", error.response.status ,"]")
+		if (error.response.data && error.response.data.msg) {
+			errorText.push(":" , error.response.data.msg,",")
```
Comment (jppiiroinen) [inline src/lib/getApiError.js]: errorText.push(":" , error.response.data.msg,",") => errorText.push(":", error.response.data.msg, ",")

File: src/lib/getApiError.js [hunk 19]
```diff
@@ -0,0 +1,21 @@
+/*
+This function returns well texted error message
+*/
+
+export default function getApiError(error,apiCall,datasetId) {
+	let errorText=[ "Error", apiCall, datasetId ]
```
Comment (jppiiroinen) [inline src/lib/getApiError.js]: let errorText=[ "Error", apiCall, datasetId ] => let errorText = [ "Error", apiCall, datasetId ]

File: src/lib/getApiError.js [hunk 20]
```diff
@@ -0,0 +1,21 @@
+/*
+This function returns well texted error message
+*/
+
+export default function getApiError(error,apiCall,datasetId) {
```
Comment (jppiiroinen) [inline src/lib/getApiError.js]: getApiError(error,apiCall,datasetId) => getApiError(error, apiCall, datasetId)

File: src/lib/getApiError.js [hunk 21]
```diff
@@ -0,0 +1,21 @@
+/*
```
Comment (jppiiroinen) [inline src/lib/getApiError.js:2]: Missing /* ADD_LICENSE_HEADER */ from the first line

**Review by jppiiroinen — CHANGES_REQUESTED** [review]

File: src/lib/getApiError.js (hunk 21 above)
Comment (ShreyasDeshpande) [reply src/lib/getApiError.js:2]: Added

File: src/lib/getApiError.js [hunk 22]
```diff
@@ -0,0 +1,22 @@
...
+	let errorText = [ "Error", apiCall, datasetId ]
+	if (error.response) {
+		errorText.push("[",error.response.status,"]")
+		if (error.response.data && error.response.data.msg) {
+			errorText.push(":",error.response.data.msg,",")
+		}
+		if (error.response.data && error.response.data.error_id) {
+			errorText.push("Error id =",error.response.data.error_id,"Please contact servicedesk(at)csc.fi with this error Id")
+		}
+	} else if (error.code  && error.code === 'ECONNABORTED') {
```
Comment (jppiiroinen) [inline src/lib/getApiError.js]: "  && " => " && "

File: src/lib/getApiError.js [hunk 23]
```diff
@@ -0,0 +1,22 @@
...
+		errorText.push(":","Request is taking too long")
+	}else if (error.message) {
```
Comment (jppiiroinen) [inline src/lib/getApiError.js]: "}else" => "} else"

File: src/views/Editor.vue (hunk 5 above)
Comment (jppiiroinen) [reply src/views/Editor.vue]: The point for this comment is that this variable has never a method assigned to it. It is just a string variable.

**Comment by jppiiroinen** [comment]
resolve merge conflicts and I will do another review check round.

**Review by jppiiroinen — APPROVED** [review]

**Pull request merged**